presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -963,7 +963,9 @@ else if (criteria instanceof JoinOn) {
                 // we do it further down when after we determine which subexpressions apply to left vs right tuple)
                 ExpressionAnalyzer analyzer = ExpressionAnalyzer.create(analysis, session, metadata, sqlParser, accessControl);
 
-                Type clauseType = analyzer.analyze(expression, output);
+                // need to register coercions in case when join criteria requires coercion (e.g. join on char(1) = char(2))
+                ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, output);
```

Comment:
But then it seems that coercions from `canonicalized` expression are not added to `Analysis` (coercions were previously added for `optimizedExpression` in https://github.com/prestodb/presto/pull/7529/commits/def126c9b42ff1416ecf2684fbeb03514a753688#diff-1a785e48f45c659076f1e4ec089517deL1004). This works until `canonicalizeExpression` doesn't change anything and returns unmodified, original expression.

Can you add a test where `canonicalizeExpression` actually changes expression and coercions are required? Maybe:
```
SELECT ... JOIN ON CAST((CASE WHEN (TRUE IS NOT NULL) THEN 'a' ELSE 'a'  END) as char(1)) = CAST('a' as char(2))
```
`TRUE IS NOT NULL` should be rewritten by `canonicalizeExpression` causing tree upward to be rewritten as well. I think it will fail with `expression not analyzed` exception from `com.facebook.presto.sql.analyzer.Analysis#getTypeWithCoercions`